File: android/app/src/main/java/com/furnit/android/utils/YoloRatioCalibration.kt
package com.furnit.android.utils

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Rect
import com.furnit.android.DetectionResult
import com.furnit.android.services.NcnnYoloe
import java.io.File
import kotlin.math.max

/**
 * Ratio-only calibration from YOLOe boxes (no metric units): wall / furniture height as
 * fraction of image height, plus helpers to crop live frames toward a target fraction.
 *
 * Kept in lockstep with iOS [Furnit/Services/OnDevice/YoloRatioCalibration.swift].
 */
object YoloRatioCalibration {

    /** Core furniture categories used for per-class target height fractions (matches iOS). */
    val FURNITURE_CANONICAL_LABELS: Set<String> = setOf(
        "chair", "couch", "bed", "dining table", "table", "wardrobe", "desk", "bench", "ottoman",
    )

    @Deprecated("Use FURNITURE_CANONICAL_LABELS", ReplaceWith("FURNITURE_CANONICAL_LABELS"))
    val FURNITURE_LABELS: Set<String> = FURNITURE_CANONICAL_LABELS

    /**
     * Map raw COCO / UI labels to canonical keys stored in metadata (matches iOS `canonicalLabel`).
     */
    fun canonicalLabel(raw: String): String {
        val trimmed = raw.trim().lowercase()
        if (trimmed.isEmpty()) return "unknown"
        if (trimmed.contains("sofa") || trimmed == "couch") return "couch"
        if (trimmed.contains("chair")) return "chair"
        if (trimmed.contains("bed")) return "bed"
        if (trimmed.contains("dining") && trimmed.contains("table")) return "dining table"
        if (trimmed.contains("wardrobe") || trimmed.contains("closet")) return "wardrobe"
        if (trimmed.contains("desk")) return "desk"
        if (trimmed.contains("ottoman")) return "ottoman"
        if (trimmed.contains("bench")) return "bench"
        if (trimmed.contains("table")) return "table"
        return trimmed
    }

    @Deprecated("Use canonicalLabel", ReplaceWith("canonicalLabel(raw)"))
    fun canonicalFurnitureLabel(raw: String): String = canonicalLabel(raw)

    /**
     * Wall-like boxes: wide vs height, moderate vertical extent, often near horizontal edges.
     * Matches iOS `wallHeightFractionOrFullFrame(imageSize:boxes:)` (not the tier-3 wall-strip heuristic).
     */
    fun wallHeightFractionOrFullFrame(
        imageWidth: Int,
        imageHeight: Int,
        boxes: List<CalibrationBox>,
    ): Float {
        val iw = max(imageWidth, 1).toFloat()
        val ih = max(imageHeight, 1).toFloat()

        var best: CalibrationBox? = null
        var bestArea = 0f

        for (box in boxes) {
            val labelCanon = canonicalLabel(box.label)
            if (labelCanon in FURNITURE_CANONICAL_LABELS) continue

            val w = box.width
            val h = box.height
            if (w <= 1f || h <= 1f) continue

            val widthFrac = w / iw
            val heightFrac = h / ih
            val aspect = w / max(h, 1f)

            val nearEdge =
                (box.centerX - w * 0.5f) / iw < 0.08f ||
                    (box.centerX + w * 0.5f) / iw > 0.92f
            val wideEnough = widthFrac > 0.35f
            val moderateHeight = heightFrac > 0.08f && heightFrac < 0.85f
            val flatEnough = aspect > 2.0f

            if (wideEnough && moderateHeight && flatEnough && (nearEdge || widthFrac > 0.55f)) {
                val area = w * h
                if (area > bestArea) {
                    bestArea = area
                    best = box
                }
            }
        }

        best?.let { wallBox ->
            val frac = (wallBox.height / ih).coerceIn(0.02f, 1f)
            return frac.coerceAtMost(1f)
        }
        return 1f
    }

    /**
     * Per canonical furniture label, 75th percentile of bbox height / image height (matches iOS).
     */
    fun furnitureHeightFractionsByLabel(
        imageHeight: Int,
        boxes: List<CalibrationBox>,
    ): Map<String, Float> {
        if (imageHeight <= 0) return emptyMap()
        val imageHeightSafe = max(imageHeight, 1).toFloat()
        val byLabel = mutableMapOf<String, MutableList<Float>>()
        for (box in boxes) {
            val canon = canonicalLabel(box.label)
            if (canon !in FURNITURE_CANONICAL_LABELS) continue
            val frac = (box.height / imageHeightSafe).coerceIn(0.01f, 0.95f)
            byLabel.getOrPut(canon) { mutableListOf() }.add(frac)
        }
        val out = mutableMapOf<String, Float>()
        for ((label, values) in byLabel) {
            if (values.isEmpty()) continue
            val sorted = values.sorted()
            // Same index as Swift: min(count-1, Int(Double(count)*0.75 - 0.001))
            val idx = kotlin.math.min(
                sorted.size - 1,
                (sorted.size * 0.75 - 0.001).toInt().coerceAtLeast(0),
            )
            out[label] = sorted[idx]
        }
        return out
    }

    /**
     * Pick a reference image for offline ratio capture (matches list thumbnails when possible).
     */
    fun pickReferenceImageFile(roomFolder: File): File? {
        if (!roomFolder.isDirectory) return null
        val candidates = listOf(
            File(roomFolder, "thumbnail.png"),
            File(roomFolder, "front_wall.png"),
            File(roomFolder, "room.jpeg"),
            File(roomFolder, "room.jpg"),
            File(roomFolder, "photo.jpg"),
        )
        return candidates.firstOrNull { it.isFile && it.length() > 0L }
    }

    fun decodeBitmapReference(file: File, maxSide: Int = 1280): Bitmap? {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(file.absolutePath, options)
        var sample = 1
        val maxDim = max(options.outWidth, options.outHeight)
        while (maxDim / sample > maxSide) sample *= 2
        val decodeOptions = BitmapFactory.Options().apply { inSampleSize = sample.coerceAtLeast(1) }
        return BitmapFactory.decodeFile(file.absolutePath, decodeOptions)?.let { bmp ->
            if (bmp.config != Bitmap.Config.ARGB_8888) bmp.copy(Bitmap.Config.ARGB_8888, false) else bmp
        }
    }

    /**
     * Target height fraction for a detection label; falls back to [defaultFrac].
     */
    fun targetHeightFracForLabel(
        label: String,
        storedByClass: Map<String, Float>,
        defaultFrac: Float,
    ): Float {
        val canon = canonicalLabel(label)
        storedByClass[canon]?.let { return it.coerceIn(0.06f, 0.85f) }
        if (canon == "couch") storedByClass["sofa"]?.let { return it.coerceIn(0.06f, 0.85f) }
        return defaultFrac.coerceIn(0.06f, 0.85f)
    }

    /**
     * Crop rect so that [bboxHeightPx] / cropHeight ≈ [targetHeightFrac] (clamped), centered on (cx, cy).
     */
    fun cropRectForTargetFurnitureFrac(
        frameWidth: Int,
        frameHeight: Int,
        centerXPx: Float,
        centerYPx: Float,
        bboxWidthPx: Float,
        bboxHeightPx: Float,
        targetHeightFrac: Float,
    ): Rect {
        val rTarget = targetHeightFrac.coerceIn(0.08f, 0.55f)
        val minCropH = (bboxHeightPx * 1.06f).toInt().coerceAtLeast(1)
        // Desired crop height so bbox occupies ~rTarget of vertical FOV in the crop
        var cropH = (bboxHeightPx / rTarget).toInt().coerceIn(minCropH, frameHeight)
        var cropW = max(bboxWidthPx * 2.1f, cropH * 0.88f).toInt()
            .coerceIn((bboxWidthPx * 1.25f).toInt().coerceAtLeast(1), frameWidth)

        var left = (centerXPx - cropW / 2f).toInt()
        var top = (centerYPx - cropH / 2f).toInt()
        left = left.coerceIn(0, max(0, frameWidth - cropW))
        top = top.coerceIn(0, max(0, frameHeight - cropH))
        cropW = cropW.coerceAtMost(frameWidth - left)
        cropH = cropH.coerceAtMost(frameHeight - top)
        return Rect(left, top, left + cropW, top + cropH)
    }

    /**
     * Map pixel-space center bbox to a square [modelSize] space (matches overlay / ONNX convention).
     */
    fun detectionResultInModelSquare(
        centerX: Float,
        centerY: Float,
        widthPx: Float,
        heightPx: Float,
        frameWidth: Int,
        frameHeight: Int,
        modelSize: Int,
        confidence: Float,
        label: String,
    ): DetectionResult {
        val basis = max(frameWidth, frameHeight).toFloat().coerceAtLeast(1f)
        val s = modelSize / basis
        return DetectionResult(
            x = centerX * s,
            y = centerY * s,
            w = widthPx * s,
            h = heightPx * s,
            confidence = confidence,
            label = label,
        )
    }

    /**
     * Paste a smaller RGBA bitmap into a transparent full-frame buffer at [left],[top].
     */
    fun pasteMaskOntoFullFrame(
        fullWidth: Int,
        fullHeight: Int,
        croppedMask: Bitmap,
        left: Int,
        top: Int,
    ): Bitmap {
        val out = Bitmap.createBitmap(fullWidth, fullHeight, Bitmap.Config.ARGB_8888)
        val canvas = android.graphics.Canvas(out)
        canvas.drawBitmap(croppedMask, left.toFloat(), top.toFloat(), null)
        return out
    }

    data class CalibrationBox(
        val label: String,
        val centerX: Float,
        val centerY: Float,
        val width: Float,
        val height: Float,
        val confidence: Float,
    )

    fun fromNcnnDetections(detections: List<NcnnYoloe.Detection>): List<CalibrationBox> =
        detections.map {
            CalibrationBox(
                label = it.label,
                centerX = it.x,
                centerY = it.y,
                width = it.width,
                height = it.height,
                confidence = it.confidence,
            )
        }
}

File: Furnit/Services/OnDevice/YoloRatioCalibration.swift
import Foundation
import CoreGraphics

// MARK: - Calibration box (image-space, pixel units)

/// Bounding box from YOLO in **image pixel coordinates** (origin top-left), center-size form.
public struct YoloCalibrationBox: Sendable {
    public var label: String
    public var centerX: CGFloat
    public var centerY: CGFloat
    public var width: CGFloat
    public var height: CGFloat
    public var confidence: Float

    public init(label: String, centerX: CGFloat, centerY: CGFloat, width: CGFloat, height: CGFloat, confidence: Float) {
        self.label = label
        self.centerX = centerX
        self.centerY = centerY
        self.width = width
        self.height = height
        self.confidence = confidence
    }
}

// MARK: - Ratio calibration (pure math, no CoreML)

/// Dimensionless YOLO height ratios for saved-room FurnitureFit (mirrors Android `YoloRatioCalibration`).
public enum YoloRatioCalibration {

    /// Core furniture categories used for per-class target height fractions.
    public static let furnitureCanonicalLabels: Set<String> = [
        "chair", "couch", "bed", "dining table", "table", "wardrobe", "desk", "bench", "ottoman"
    ]

    /// Map raw COCO / UI labels to canonical keys stored in metadata.
    public static func canonicalLabel(raw: String) -> String {
        let trimmed = raw.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return "unknown" }

        // Synonyms → canonical
        if trimmed.contains("sofa") || trimmed == "couch" { return "couch" }
        if trimmed.contains("chair") { return "chair" }
        if trimmed.contains("bed") { return "bed" }
        if trimmed.contains("dining") && trimmed.contains("table") { return "dining table" }
        if trimmed.contains("wardrobe") || trimmed.contains("closet") { return "wardrobe" }
        if trimmed.contains("desk") { return "desk" }
        if trimmed.contains("ottoman") { return "ottoman" }
        if trimmed.contains("bench") { return "bench" }
        if trimmed.contains("table") { return "table" }

        return trimmed
    }

    /// Wall-like boxes: very wide vs height, moderate vertical extent, often near horizontal edges.
    public static func wallHeightFractionOrFullFrame(imageSize: CGSize, boxes: [YoloCalibrationBox]) -> Float {
        let imageWidth = max(imageSize.width, 1)
        let imageHeight = max(imageSize.height, 1)

        var best: YoloCalibrationBox?
        var bestArea: CGFloat = 0

        for box in boxes {
            let labelCanon = canonicalLabel(raw: box.label)
            if furnitureCanonicalLabels.contains(labelCanon) { continue }

            let w = box.width
            let h = box.height
            guard w > 1, h > 1 else { continue }

            let widthFrac = Float(w / imageWidth)
            let heightFrac = Float(h / imageHeight)
            let aspect = w / max(h, 1)

            // Heuristic: wall segments are wide and not extremely tall
            let nearEdge = (box.centerX - w * 0.5) / imageWidth < 0.08 ||
                (box.centerX + w * 0.5) / imageWidth > 0.92
            let wideEnough = widthFrac > 0.35
            let moderateHeight = heightFrac > 0.08 && heightFrac < 0.85
            let flatEnough = aspect > 2.0

            if wideEnough && moderateHeight && flatEnough && (nearEdge || widthFrac > 0.55) {
                let area = w * h
                if area > bestArea {
                    bestArea = area
                    best = box
                }
            }
        }

        if let wallBox = best {
            let frac = Float(wallBox.height / imageHeight)
            return min(1.0, max(0.02, frac))
        }
        return 1.0
    }

    /// Per canonical furniture label, robust height / imageHeight in [0.01, 0.95].
    public static func furnitureHeightFractionsByLabel(imageHeight: CGFloat, boxes: [YoloCalibrationBox]) -> [String: Float] {
        let imageHeightSafe = max(imageHeight, 1)
        var byLabel: [String: [Float]] = [:]

        for box in boxes {
            let canon = canonicalLabel(raw: box.label)
            guard furnitureCanonicalLabels.contains(canon) else { continue }
            let frac = Float(box.height / imageHeightSafe)
            let clamped = min(0.95, max(0.01, frac))
            byLabel[canon, default: []].append(clamped)
        }

        var result: [String: Float] = [:]
        for (label, values) in byLabel {
            let sorted = values.sorted()
            guard !sorted.isEmpty else { continue }
            let idx = min(sorted.count - 1, Int(Double(sorted.count) * 0.75 - 0.001))
            result[label] = sorted[idx]
        }
        return result
    }
}

// MARK: - ROI geometry for live FurnitureFit second pass

public enum FurnitureFitRatioGeometry {

    /// Crop rect in **pixel buffer coordinates** (origin top-left) so that bbox height / cropHeight ≈ `rTarget`.
    public static func cropRectForTargetFurnitureFraction(
        frameWidth: Int,
        frameHeight: Int,
        bboxMinX: CGFloat,
        bboxMinY: CGFloat,
        bboxMaxX: CGFloat,
        bboxMaxY: CGFloat,
        rTarget: Float,
        minCropAreaFraction: Float = 0.12,
        maxCropAreaFraction: Float = 0.92
    ) -> CGRect? {
        guard frameWidth > 8, frameHeight > 8 else { return nil }
        let fw = CGFloat(frameWidth)
        let fh = CGFloat(frameHeight)

        let bh = max(1, bboxMaxY - bboxMinY)
        let cx = (bboxMinX + bboxMaxX) * 0.5
        let cy = (bboxMinY + bboxMaxY) * 0.5

        let rt = CGFloat(max(0.05, min(0.9, rTarget)))
        // Desired crop height so bh / cropH ≈ rt → cropH ≈ bh / rt
        var cropH = min(fh, max(bh * 1.15 / rt, bh * 1.25))
        cropH = min(cropH, fh)
        var cropW = min(fw, cropH * (fw / fh))
        cropH = min(cropH, fh)
        cropW = min(cropW, fw)

        let areaFrac = Float((cropW * cropH) / (fw * fh))

        // Reject only *too small* crops; for very large objects where the ideal crop
        // would exceed `maxCropAreaFraction`, fall back to using (almost) the whole
        // frame instead of aborting the ratio pass. This matches the UX expectation
        // of “zooming out as much as the camera allows” rather than silently
        // disabling the second pass.
        if areaFrac < minCropAreaFraction {
            return nil
        }

        let rect: CGRect
        if areaFrac > maxCropAreaFraction {
            // Fallback: ratio target is not achievable without exceeding the frame.
            // Use the full frame so downstream code can still run a second pass and
            // log what happened, even though r_curr will remain larger than r_target.
            rect = CGRect(x: 0, y: 0, width: fw, height: fh)
        } else {
            var originX = cx - cropW * 0.5
            var originY = cy - cropH * 0.5
            originX = max(0, min(fw - cropW, originX))
            originY = max(0, min(fh - cropH, originY))
            rect = CGRect(x: originX, y: originY, width: cropW, height: cropH)
        }

        if rect.width < 32 || rect.height < 32 { return nil }
        return rect
    }

    /// After second pass, verify height fraction in crop; used for debugging / optional rejection.
    public static func heightFractionInCrop(bboxHeightPx: CGFloat, cropHeightPx: CGFloat) -> Float {
        let ch = max(cropHeightPx, 1)
        return Float(min(1, max(0, bboxHeightPx / ch)))
    }
}
